### DISCUSSION TITLE
Power & Damage Control

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#229] [Feature Proposals]
_For more information about the ship systems discussions, see https://github.com/Thorium-Sim/thorium-nova/discussions/224_

# Power & Damage Control

Ship systems require power and many systems generate heat.

### Power

In Thorium Classic, power is generated by a single Reactor, and can be stored in batteries. Power is represented in discrete units (though the amount is unitless, and doesn't correspond to real-world units like kilowatts) and distributed to systems in those units. The reactor can be set to different "efficiencies" which increases or decreases its output and heat usage accordingly. This is done to increase stealthiness or as part of a damage report.

Each ship system required a certain number of power units to operate, and additional units could be supplied to increase the capabilities of the system. More power in the warp engines meant higher warp speeds, additional tractor beam power meant stronger tractor beams, etc. A system will stop working entirely if power is removed from it.

If the ship has no batteries, the intention is that the power usage must match the power generated by the reactor exactly. If there are batteries, using less power charges the batteries with the excess, and using more power drains the batteries. There is no built-in consequence for not having power balanced, or if the batteries become drained - the Flight Director gets to deal with that.

Heat has no correlation with power in Thorium Classic and is largely arbitrary in how fast it is generated. Systems heat up as they are used and slowly cool down when they are not in use. Coolant, an expendable resource, can be flushed through the systems to immediately cool them down. There is also no built-in consequence for overheating - the Flight Director is responsible for noticing and warning the crew, or damaging a system.

### Damage Control

In Thorium Classic, systems are either operational, or damaged. Systems are only damaged at the whim of the Flight Director or through timeline events. When a system is damaged, it's controls are no longer usable by the crew member. 

Repairing the damaged system requires following a number of steps. In the old text-based damage reports, those steps were automatically generated but the system didn't track whether the steps were completed or not. At the end of the report, a reactivation code is entered which indicates to the Flight Director that the damage report was complete.

The new task-based damage reports allowed the steps in the reports to be tracked by Thorium Classic to know when they were completed. Once the step is completed, the damage report progresses to the next step. The Flight Director is notified when the report is complete, and the system is repaired.

## Thorium Nova

The hope with Thorium Nova is to improve the realism and the depth of gameplay in both power and damage control. One way we can do this is by linking the two concepts.

### Efficiency

Instead of merely being damaged or not damaged, systems have an "efficiency" level which affects how well they work. For the most part, lower efficiency means a system requires more power to operate and generates more heat. Systems that consume other resource perhaps require more resources to operate. And systems that can have some kind of delay in their operation, like Long Range Communications, take even longer to perform their tasks.

Efficiency is essentially a ratio. Multiplying the power provided by that ratio yields the actual work that the system can do. However, heat generation, which is a factor of power usage, either stays the same or increases with the decreased efficiency.

I believe that it makes the most sense to model damage and efficiency separately within each system, where damage is a percentage and the efficiency is calculated from the damage, heat and any other factors that affect efficiency. 

Damage can increase in one of four ways:
- Entropy, which slowly increases the damage to the system, decreasing efficiency.
- Heat - Heat affects efficiency in two ways. First, the heat of a system will directly impact the efficiency of the system. Cooling the system will immediately improve efficiency. Second, the longer a system overheats, the more damaged it becomes, which also decreases the efficiency. This can only be resolved by performing damage control tasks.
- Sabotage - A disgruntled crew member or intruder could purposely damage a system, decreasing its efficiency.
- Space combat - Getting hit with weapons will cause damage and decrease the efficiency of ship systems.
  
### Damage Control

Damage Control then becomes a role of maintenance - choosing systems that the crew might need which are low efficiency due to high damage, and performing tasks to decrease their damage and increase their efficiency.

Naturally, it should still be possible to have systems be completely inoperable until they are repaired. Damaging the system 100% would do that.

Naturally, Damage Control should be a fully automated process, where steps to repair systems are automatically generated and their completion can be automatically validated. Once the steps are completed, the system's damage is decreased.

### Reactors and the Power Grid

There are a _lot_ of ways that we can model power distribution. I [wrote a lot about this](https://thoriumsim.com/blog/distributing-power) last year, and I think it could serve as a good starting point for a conversation. Be sure to read all of that - there's a lot there.

Here's the tl;dr version:
- Multiple reactors provide redundancy and novel ways to model the power grid.
- Reactors operate on a supply/demand system, where the reactor will generate more power if there is more demand.
- When a reactor's demand is decreased, the supply will instantly be reduced, but if demand increases again, it will take some time for the reactor to spin up again, and supply will take some time to reach the new demand. 
- Batteries fill the gap when a reactor is not currently producing enough power. Batteries are charged by the reactors (I suppose you can choose how much power the reactor should generate to charge the battery, which affects how fast it is charged), and are drained by systems that are using them. Batteries have a max output, so even a full battery can't be discharged very quickly.
- Power nodes in a power grid allow power to be distributed from reactors and batteries to systems. Power nodes can be damaged or lose efficiency, affecting power further up the grid. 
- If ship systems don't get as much power as they need, the operate poorly. Ship systems will low efficiency will attempt to draw more power to operate.
- Special power nodes, like the Warp Dynamo or Phaser Capacitor, can be charged up like batteries, and then discharged rapidly. They are typically used for their respective systems, but can be attached to other systems to make that system do something with more power. Not advised, since it also damages that system.

Read the blog post above for more details about all of this.

---

Am I missing anything? Comments, suggestions, feedback, or alternate proposals welcome!

**Comment by Froginator45** [comment]
Something that could be interesting is to include some battery physics. Efficiency of the batteries can be reduced over time at a rate that is proportional to the battery usage. As the efficiency is reduced, the total battery capacity is reduced. This makes the usability of the battery banks less and less. At some point, repairs might need to be made to make the ship ready for more emergency situations. The efficiency could effect each separate battery bank if there are multiple divided out to different systems. 

It might also be interesting to include the ability to include additional large capacitors that can be applied to different systems, with the obvious bonus of being able to supply larger amounts of power quickly but needing time to charge up more fully. Useful if the crew needs a big burst of energy for some system but don't want to Warp Dynamo or other similar system. I'm not entirely sure if this would be needed since it would function the same as you described the Warp Dynamo or Phaser Capacitor. It would just act more of like a primary possibility and the other systems would then be redundant in those situations.

**Reply by alexanderson1993** [reply]
Battery (and power node) efficiency and damage is a good idea.

I think the Phaser Capacitors would serve the general capacitor role well. Part of me wants to make their ability to be hooked up to any system non-obvious, so it becomes a nice Easter egg/cheat code for curious crews. Keeping them named "Warp Dynamo" and "Phaser Capacitor" helps obfuscate their general-purpose nature.

**Comment by Ryan-Anderson-PhD** [comment]
I love the idea of having ongoing entropy in systems that needs to be maintained and addressed. This is great.

I also wonder about adding the possibility for upgrades to systems. Perhaps this is done by creating parts at another station (perhaps operations or science) and installing them. These could increase maximum efficiency, slow down entropy due to heat, reduce damage, etc. Different upgrades could attend to each one of these functions for each system. So, it could be a way to reinforce the ship, and would give logical tasks for science and/or operations to do on a regular basis.

**Reply by Froginator45** [reply]
I think upgrades would be cool but mostly useful for campaigns (or if specific for the mission). As I see it, upgrades could be obtained from a store/trading, stolen from enemies in a mission, or invented by the operations or science team. The upgrades would need to be implemented/approved by engineering after they have been obtained. Inventing the device and integrating the device into the ships systems could both definitely be new cards for the crew.

**Comment by Ryan-Anderson-PhD** [comment]
It would also open up the possibility of sometimes upgrades not working or
making things worse, which could be a fun element.

Sincerely,

Ryan Anderson, Ph.D., LMFT, MedFT
Co-founder, Telos U
Executive Director, Telos Discovery Space Center
He/Him
***@***.***
Cell: 801-505-3401
https://telos.org/
discoveryspacecenter.com


On Thu, Feb 24, 2022 at 4:14 PM Froginator45 ***@***.***>
wrote:

> I think upgrades would be cool but mostly useful for campaigns (or if
> specific for the mission). As I see it, upgrades could be obtained from a
> store/trading, stolen from enemies in a mission, or invented by the
> operations or science team. The upgrades would need to be
> implemented/approved by engineering after they have been obtained.
> Inventing the device and integrating the device into the ships systems
> could both definitely be new cards for the crew.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Thorium-Sim/thorium-nova/discussions/229#discussioncomment-2247463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWAOTBYMN5VL4CKLF56SM3TU423WLANCNFSM5O6Y6OGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> .com>
>

-- 

    *Telos is a* *Proud Member of Choose Mental Health 
<https://choosementalhealth.ck.page/7de19e5d70>*

**Comment by alexanderson1993** [comment]
One feature that should be included: Being able to save power routes and switch between them quickly. That makes it possible to pre-configure different power routes for travel, battle, stealth, etc and switch between them as needed.

**Comment by 1deejay** [comment]
Damage control can also be done by sending reams to fix things, the more teams available the more you can send to fix something. 

How much raw material is available for repairs? We can look at having mining probes get raw materials from asteroids in a pinch of we can't restock at a station.
